### DISCUSSION TITLE
Farcaster Hubs as an Events Buffer for Social Applications

### DISCUSSION THREAD

**Discussion by michaelhly** [farcasterxyz/protocol#88] [General]
# Problem
Traditionally, application content and events exist within walled garden social platforms. For instance, a user’s achievements on [Strava](https://www.strava.com/) are by-default inaccessible to other social apps like Instagram and Twitter. In order to make user events or content available across applications, sharing functionality must be exposed through application-specific APIs or SDKs. Here are some examples:
- https://docs.snap.com/api/home#lens-studio
- https://developers.facebook.com/docs/instagram/sharing-to-feed/

Consuming developers must then install dependencies within their applications to enable users to publish content to each social platform. These SDKs may be poorly maintained [[1](https://github.com/Snapchat/snap-kit-spm/issues/10)][[2](https://github.com/keybase/client/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc)], and in more extreme cases, these APIs can be [shut off](https://www.engadget.com/twitter-shut-off-its-free-api-and-its-breaking-a-lot-of-apps-222011637.html) completely and break the developer’s application.

Farcaster renders this frustrating experience unnecessary; application events and content can be published and become easily accessible from Hubs. 

However, a problem exists within the Farcaster ecosystem today. Most applications in the Farcaster ecosystem are first built with Warpcast APIs instead of Hub APIs. Warpcast APIs are very opinionated and specifically designed toward supporting a Twitter-like application and primarily serves a community of crypto/web3 enthusiasts. Applications in the Farcaster ecosystem all share one **primary** feature — allow users to publish tweet-like data to a ~~primary/master~~ distributed data store and surface tweet-like content based on differing client logic.

# Motivation
To facilitate an ecosystem of diverse applications, applications must migrate off of Warpcast APIs. But to a further extent, the Farcaster “protocol” should be less opinionated and allow application developers to focus on developing “social” experiences for their users. For example, application developers should have the degree of freedom to choose if:
- users need to be onboarded to a blockchain
- cryptocurrency *needs* to be used to handle payments
- content from application X should be surfaced on application Y

Stewards of the Farcaster protocol should focus on alleviating developers from writing + maintaining application-specific external APIs and SDKs. Farcaster, the protocol, should start off functioning more like a generic events buffer without imposing social primitives onto applications. Applications should also be encouraged to publish all kinds of content to Farcaster Hubs without affecting the user experience of another application. 

Clients who wish to support content published by certain applications should read from hubs, verify that the published message was signed by an application against a cryptographic proof, and surface the message in their application appropriately.

Example:
1. A user wants to share the completion of his/her daily steps goal from [Blobs](https://blobs.lol/) on Warpcast
2. Blobs would submit a message to Hubs on behalf of the user via [signer requests](https://warpcast.notion.site/Warpcast-v2-API-Documentation-c19a9494383a4ce0bd28db6d44d99ea8#13cb11cba1114002adc928c5fa679364)
3. Message submitted by Blobs should include a signed + verifiable signature (can be in canonical URI form or a new kind of proto message)
4. Warpcast reads an application-specific message from hubs
5. Warpcast verifies that the message was published from a known/supported client
6. Warpcast renders the message as a cast, sticker, badge, etc. on the user’s profile

Warpcast can also decide it does not like messages produced by the Blobs application and by default opt out of messages generated by Blobs and other unknown applications.

# Proposal
1. Create an on-chain registry of known applications 
- Applications should publish some sort of verifiable proof to this registry where other applications could verify the integrity of the message published by a known/supported application.
- Applications pay a yearly fee to maintain their known status in the registry 

~~2. Move FIDs off-chain~~
~~- User identifiers should be moved off-chain but applications should still agree on a canonical standard for user identification~~

**Comment by michaelhly** [comment]
Another pro to this approach is that the Application registry can be deployed to all kinds of L1s, L2s, or some future implementation of some distributed ledger without forcing a user migration.

**Comment by michaelhly** [comment]
Also, bots should be registered as applications, and distinguished from users.

**Reply by varunsrin** [reply]
a self-attestation for "automated account" would be a nice addition. is this what you were thinking of in terms of a solution?

**Comment by varunsrin** [comment]
Thanks for writing this up - lots of great suggestions in here with different tradeoffs, so let me try to break them down and address them individually. 

> Create an on-chain registry of known applications

This is a great idea, and there are two ways we can do this. One is to have a distinct on-chain registry and the other is to just require each app to have an fid and a special message in its USER_DATA identifying it as an application. The former gives us a lot more flexibility in terms of charging fees the latter is simple and composable with the existing system. 

> Applications should publish some sort of verifiable proof to this registry where other applications could verify the integrity of the message published by a known/supported application.

Doing this at the message level might be tedious (2-way signing is just annoying in practice) but you have apps and users both co-sign the creation of a signer, and then you can assume that all messages from that signer came from the app and the user. 

> 2. Move FIDs off-chain

> Another pro to this approach is that the Application registry can be deployed to all kinds of L1s, L2s, or some future implementation of some distributed ledger without forcing a user migration.

We've thought a lot about this, and its quite hard to do in practice. The purely off-chain approach doesn't work because how does a user know they can move an identify from one app to another? 

The multi-chain approach is appealing, but now applications must sync with N different blockchains to know the actual state of users. This is actually quite tedious in practice and there are lot of state management complexities both for hubs and applications. A much better approach is to have a singleton registry on a blockchain.

**Reply by michaelhly** [reply]
Perhaps I didn't distill my ideas well in the proposal part, so I want to try again.

Here is what I see today: Discove, Jam, and Warpcast are all building Twitter-like web and mobile clients over Farcaster. It seems like a lot of resources are being used to deliver the same user experience but in different flavors.

I think based on how Hubs is designed, it seems more advantageous for the protocol to serve different specialized clients across different platforms. For example, we can have one team building a mobile fitness tracker, a second team building a web-based prediction market, and a third team building a VR simulation all reading/writing content to a shared events buffer (Hubs).

The on-chain registry is for applications to filter messages for specific known applications that a client wishes to include in their product. Each client would selectively filter for user-generated content published by other clients (whatever user content makes each respective product better).

tldr: Instead of having clients compete for user attention and serve the same types of user content from Farcaster, Farcaster can explore to facilitate coordination between different specialized clients, across platforms.

**Reply by varunsrin** [reply]
very supportive of that top-level goal, i've written down some thoughts about that here: https://hackmd.io/@farcasterxyz/Hyyt7NFb2

curious if you agree directionally with the proposed approaches

**Reply by nichochar** [reply]
Also greatly supportive of that top level goal, and agree that the protocol seems like it might have to work pro-actively to materialize it, given what we're observing with the current client diversity.

I think that your thinking is going in the right direction, and can be pushed.
The thought exercise is: What would the protocol need to provide (primitives-wise) to _easily_ build:
- an instagram clone
- a tiktok clone
- a strava clone
- a twitter clone
- a facebook groups clone

I am pretty convinced that the vertex between two nodes (a node is a user, a vertex is a relationship, like "follow" or "amp"), is _the_ key thing to get right, but I don't have a clear picture on what the solution space looks like for alternatives that support this.

@varunsrin your article is right on the money. I got excited reading it!
Some reactions:
> Flexible schemas

I prefer "extendible" schemas that are simple, rather than flexible.

>Long-term storage

I think that it's one of the options but an another is a data referencing protocol, which decouples the storage from the protocol entirely.

>Private casting so that developers can design private networks

I have some more thoughts about this, and working on something.

**Reply by michaelhly** [reply]
@varunsrin 
> We've thought a lot about this, and its quite hard to do in practice. The purely off-chain approach doesn't work because how does a user know they can move an identify from one app to another?

How about a system where the protocol auctions off blocks of FIDs to known apps to distribute? Like how ICANN or regional internet registries assign IP address blocks. Can elaborate on implementation if this sounds interesting ...

**Reply by varunsrin** [reply]
High level response is that if we make FID registration open on L1 (which we plan to by fall) then the auctioning blocks seems unnecessary? Anyone can just register directly with the blockchain. 

If I'm misunderstanding the proposal, an elaboration would be helpful

**Reply by michaelhly** [reply]
Social applications require [existing users to invite new users](https://en.wikipedia.org/wiki/K-factor_(marketing)) for growth. Asking a new user to pay for something upfront is a non-trivial deterrent for growth. A lot of times, bursts of sign-ups can be attributed to momentary viral events and targeted marketing campaigns ... during these moments if it takes up to $20 for users to register an FID to onboard (variance due to gas price fluctuation), this will spell doom for a social application that might have worked. Also, popular consumer applications like Uber, Coinbase, etc. all implement growth strategies providing new user incentives. To date, I have yet to find a social app requiring paid-sign ups to reach meaningful network effects, so putting the cost of FID registration on the user seems very counterintuitive to me.

Based on some quick stats from https://farcaster.network/, we can see that out of ~13k sign-ups, there are about ~3000 monthly active users. So around 1 out of 4 users who signed up to Warpcast really need an FID. Instead of having "anyone" register for FIDs on an L1, we should allow applications to purchase FIDs in blocks, and abstract away FID registration in a manner where
- cost can be more predictable
- FID distribution is based on actual usage rates
- FIDs can be recycled from churned users to new users

If we allow applications to bid on blocks of FIDs, applications can come up with their own FID distribution strategies. FID registration can be handled in the following ways:
- onboard users like Warpcast, mandate users to register for an FID at signup (user pays application)
- onboard users and register FIDs later only for recurring users who have registered an Ethereum address (user bears zero cost)
- onboard users and assign FIDs ephemerally, prompt users to purchase their FID or they'll lose X, Y, and Z benefits, and recycle FIDs for churned users

**Comment by nichochar** [comment]
### General feedback

I agree with the some things, but not all. I would frame it as:
How can the farcaster protocol become more generic and extendable?

I think that it might be worth spending more time on the actual requirements, and then separate out the proposals (solutions) cleanly.

### Specifics

> Farcaster, the protocol, should start off functioning more like a generic events buffer without imposing social primitives onto applications

I disagree with this. I think that farcaster at its core is a social protocol, and shouldn't move away from that. What should happen is that it should become extendable in a way that you don't necessarily post "casts", but you can post "richCasts", and there should be ways for people to permissionlessly create a new richCast type (like `blobAchievement`), which can be posted to someone's timeline.

Then, clients can chose which types they filter. You could have an app that filters only blob stuff, one that mixes text events and strava-like events, etc...

Before this can happen though, I think there needs to be a better answer around the following graph, and if this is scoped to an application (like the ones you are thinking about with your registry) or global. 

>Create an on-chain registry of known applications

This sounds interesting and I need to think more about it.

**Reply by michaelhly** [reply]
Thank you for your feedback! I do realize the framing was a bit sporadic, so I gave it another try here:
https://github.com/farcasterxyz/protocol/discussions/88#discussioncomment-5738301

Based on the discussions here https://github.com/farcasterxyz/protocol/discussions/85. It seems too early to decide on which social primitive the protocol should enforce:
- Amps seem unproven but are a more natural fit for the technical design of the protocol
- Follows are proven, but kind of awkward for a message-graph-based protocol

(not on the team, my two cents aren't worth much here)

Wanted to offer a different perspective on how we can look at the protocol, and perhaps come up with designs to place the protocol to support applications better :)

**Comment by varunsrin** [comment]
Application identifiers were proposed and merged as part of FIP-7, which addresses a large part of this proposal: https://github.com/farcasterxyz/protocol/discussions/103

Thanks for starting the discussion around this @michaelhly!
